judge/ratings.py

```diff
@@ -28,94 +67,77 @@ def tie_ranker(iterable, key=attrgetter('points')):
         yield rank + (delta - 1) / 2.0
 
 
-def rational_approximation(t):
-    # Abramowitz and Stegun formula 26.2.23.
-    # The absolute value of the error should be less than 4.5 e-4.
-    c = [2.515517, 0.802853, 0.010328]
-    d = [1.432788, 0.189269, 0.001308]
-    numerator = (c[2] * t + c[1]) * t + c[0]
-    denominator = ((d[2] * t + d[1]) * t + d[0]) * t + 1.0
-    return t - numerator / denominator
-
-
-def normal_CDF_inverse(p):
-    assert 0.0 < p < 1
-
-    # See article above for explanation of this section.
-    if p < 0.5:
-        # F^-1(p) = - G^-1(p)
-        return -rational_approximation(math.sqrt(-2.0 * math.log(p)))
+def get_var(times_ranked, cache):
+    while times_ranked >= len(cache):
+        next_var = 1. / (1. / (cache[-1] + VAR_PER_CONTEST) + 1. / BETA2)
+        cache.append(next_var)
+    return cache[times_ranked]
+
+
+def recalculate_ratings(ranking, old_mean, times_ranked, historical_p):
+    n = len(ranking)
+    cache = [VAR_INIT]
+
+    delta = [(get_var(t, cache) + VAR_PER_CONTEST + BETA2) ** .5 for t in times_ranked]
+
+    new_p = [0.] * n
+    new_mean = [0.] * n
+
+    # Calculate performance at index i.
+    def solve_idx(i, bounds=VALID_RANGE):
+        r = ranking[i]
+        tanh_terms = []
+        y_tg = 0
+        for j, s in enumerate(ranking):
+            tanh_terms.append((old_mean[j], delta[j] * TANH_C, 1))
+            if s > r:       # j loses to i
+                y_tg += 1. / (TANH_C * delta[j])
+            elif s < r:     # j beats i
+                y_tg -= 1. / (TANH_C * delta[j])
+            # Otherwise, this is a tie that counts as half a win, as per Elo-MMR.
+        new_p[i] = solve(tanh_terms, y_tg, bounds=bounds)
+        historical_p[i] = [new_p[i]] + historical_p[i]
```

Comment:
Might be worth considering a `deque` if prepends happen a lot. Although `deque` has its own overhead, so it's probably worth benchmarking.

Reply:
I moved this code to where it's needed (see `Make intent of historical_p clearer`).
Each list is involved in 1 prepend, so the `deque` overhead may not be worth it.